src/GafferDispatch/Dispatcher.cpp

```diff
@@ -661,11 +681,12 @@ void Dispatcher::dispatch( const std::vector<NodePtr> &nodes ) const
 		}
 	}
 
-	Context::EditableScope jobScope( Context::current() );
 	// create the job directory now, so it's available in preDispatchSignal().
 	/// \todo: move directory creation between preDispatchSignal() and dispatchSignal()
-	m_jobDirectory = createJobDirectory( Context::current() );
-	jobScope.set( g_jobDirectoryContextEntry, m_jobDirectory );
+
+	ContextPtr jobContext = new Context( *Context::current() );
+	Context::Scope jobScope( jobContext.get() );
+	createJobDirectory( script, jobContext.get() );
```

Comment:
Good catch - any thoughts on how you'd prefer this to work then Ivan? I could move the responsibility for saving the file back to `_doDispatch()`, but that feels a bit like a backward step, given that all the other logic for directory management is here in the base class. Alternatively, we could make directory creation lazy, so the base class creates it and saves the script on request, with the request typically coming from `_doDispatch()`. But that will bounce us back to the race condition discussion we had over email. Might be a good topic for discussion in tomorrow's meeting?

Reply:
Can't we just move the saving of the script from the `createJobDirectory()` to just prior to calling `doDispatch()` (around line 742)?
That way the dispatch events as well as the immediate tasks would have been processed.

Reply:
Good idea - nice and simple. I've done that in d9862e0, and added some comments about how this might be improved in the future. I am a bit worried that the dispatch sequence isn't entirely logical at present, so when time allows it'd be good to chat about how it might be improved while still allowing Jabuka to do what it needs to.

Reply:
Sounds good. 

Just as a note, the reason we need to perform those operations during the immediate pre tasks is that we don't have a mechanism in Gaffer to pass dependency information from TaskNodes at the moment, which forces us to pre-create all the necessary versions and alter the graph in order to store the information about those versions for downstream nodes.

We have recently implemented a mechanism to save information to be used by TaskNodes by saving data to an extra file in the job directory, which now allows us to create the versions only when the TaskNode gets executed. By using it, we could in theory update all our publishing nodes to not require the script to be altered during dispatch.

But it would be better if Gaffer provided a similar mechanism natively.

Reply:
Cool - that gives me some additional motivation to get information-passing between TaskNodes working on the Gaffer side.